### DISCUSSION TITLE
pypdf will drop Python 3.6 support in January 2024

### DISCUSSION THREAD

**Discussion by MartinThoma** [py-pdf/pypdf#2231] [Announcements]
pypdf will drop Python 3.6 support with the first release in 2024.

The reasons for dropping support are:

* **End of Life**: Python 3.6 reached its EOL in 2021 ([PEP 494](https://peps.python.org/pep-0494/)). That was 2 years ago.
* **Usage**: About 0.2% - 0.3% of [the pypdf downloads](https://pypistats.org/packages/pypdf) are from Python 3.6. More people with Python 2.7 download pypdf.
* **Work-around**: The work-around for people who are stuck with Python 3.6 is to pin the version of pypdf. You will not get updates and no new features, but it will continue to work.
* **Reduce work**: Removing support for Python 3.6 means (1) we can use `ruff` for all supported versions (2) one less CI run (3) a little bit of progress to use modern Python features. It's now dataclasses that we could use :tada: 

Please note: We will not prevent Python 3.6 from working just for the sake of it. It's more a "your mileage will vary" approach. We will not consider Python 3.6 anymore when we continue pypdf development.

See https://github.com/py-pdf/pypdf/discussions/2005

**Comment by cthoyt** [comment]
Why not go even further and also drop Python 3.7, which is also past end of life?

**Reply by MartinThoma** [reply]
Step by step :-)

**Comment by stefan6419846** [comment]
As always, the download stats should probably be viewed with some grain of salt, due to companies using dedicated internal mirrors/PyPI subsets, while having to rely on legacy code, for example. Nevertheless, I am open to dropping Python 3.6 support as well, although *dataclasses* are not really an argument, as there is a good drop-in backport package for it (in contrast to `importlib.metadata` and `importlib.resources` and their backports as well as the `typing_extensions` package).

**Reply by pubpub-zz** [reply]
I propose to take opportunity also to move to 4.X branch and get rid of all the depreciated code

**Reply by MartinThoma** [reply]
I agree @stefan6419846

> I propose to take opportunity also to move to 4.X branch and get rid of all the depreciated code

Yes, good point :-) I wanted to do that for a while. I'm looking forward to the January cleanup :-)
